### PR TITLE
FIX/ENH CL: Allow single parent colorbar w/ gridspec layout

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

In the master version of constrained_layout (CL), a colorbar associated with a single parent axes (i.e. `fig.colorbar(im, ax=axs[2])`) is associated with the subplotspec layout box. If there is more than one parent, it is associated with the gridspec layout box that contains the subplotspecs (i.e. `fig.colorbar(im, ax=axs[:2])`).  

As noted in #10627 by @afvincent this leads to a surprising result for a layout like:

```python
import matplotlib.pyplot as plt
import numpy as np

def example_pcolor(ax, cmap='RdBu_r'):
    dx, dy = 0.6, 0.6
    y, x = np.mgrid[slice(-3, 3 + dy, dy),
                    slice(-3, 3 + dx, dx)]
    z = (1 - x / 2. + x ** 5 + y ** 3) * np.exp(-x ** 2 - y ** 2)
    pcm = ax.pcolormesh(x, y, z, cmap=cmap, vmin=-1., vmax=1.,
                        rasterized=True)
    return pcm

fig, axs = plt.subplots(3, 1, figsize=(4, 4), constrained_layout=True)
for ax in axs[:2]:
    pc = example_pcolor(ax)
pc2 = example_pcolor(axs[2], cmap='viridis')

fig.colorbar(pc, ax=axs[:2])
fig.colorbar(pc, ax=axs[2])

plt.show()
```

yields:

figure1

This is technically correct.  We *want* to be able to have a single colorbar follow a single axes:

figure3

However, for this case it is undesirable.  This PR changes the argument of `ax` in `fig.colorbar` so that if it is an iterable, even with just one element, it will be treated as a gridspec level colorbar.  i.e. if instead of `fig.colorbar(pc, ax=axs[2])` we make the axes a list:
```python
fig.colorbar(pc, ax=[axs[2]])
```
we get:

figure2

(the user will still have to futz around w/ the colorbar aspect ratios to make that look nice).  

This doesn't break `constrained_layout=False` (default), which both yeild identical layouts:

figure4

so this is a non-breaking change...

~The todo list below is still active.  I'll add a test and update the tutorial if this approach is OK.~

Decided against a dedicated test. The underlying functionality works, so a test is a bit superfluous.  Did update the tutorial.

## PR Checklist

- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant

### CODE REVIEW

**Comment by jklymak** [comment]
New doc is: https://7023-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/tutorials/intermediate/constrainedlayout_guide.html#sphx-glr-tutorials-intermediate-constrainedlayout-guide-py

File: lib/matplotlib/axes/_subplots.py [hunk 1]
```diff
@@ -173,7 +173,9 @@ def _make_twin_axes(self, *kl, **kwargs):
         """
         from matplotlib.projections import process_projection_requirements
         if 'sharex' in kwargs and 'sharey' in kwargs:
-            raise ValueError("Twinned Axes may share only one axis.")
+            if (not (kwargs['sharex'] == self ) and
```
Comment (afvincent) [inline lib/matplotlib/axes/_subplots.py]: Extra space before the parenthesis `== self )`: PEP8 seems to be unhappy.

File: lib/matplotlib/axes/_subplots.py [hunk 2]
```diff
@@ -173,7 +173,9 @@ def _make_twin_axes(self, *kl, **kwargs):
...
+                        not (kwargs['sharex'] == self)):
+                raise ValueError("Twinned Axes may share only one axis.")
```
Comment (afvincent) [inline lib/matplotlib/axes/_subplots.py]: Are those line a trial for a 2-in-1 PR done on purpose regarding to #10585 ?

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 3]
```diff
@@ -105,7 +105,7 @@ def example_plot(ax, fontsize=12, nodec=False):
 fig.colorbar(im, ax=ax, shrink=0.6)
 
 ############################################################################
-# If you specify multiple axes to the ``ax`` argument of ``colorbar``,
+# If you specify a list of axes to the ``ax`` argument of ``colorbar``,
```
Comment (afvincent) [inline tutorials/intermediate/constrainedlayout_guide.py]: If I understood correctly, even a tuple or an ndarray (or actually any kind of iterable...) will trigger that behavior, will it not? Should this be briefly mentioned or would it make the API even harder to understand :/?

File: tutorials/intermediate/constrainedlayout_guide.py [hunk 4]
```diff
@@ -114,6 +114,31 @@ def example_plot(ax, fontsize=12, nodec=False):
...
+fig, axs = plt.subplots(3, 1, figsize=(4, 4), constrained_layout=True)
+for ax in axs[:2]:
+    im = ax.pcolormesh(arr, rasterized=True)
+fig.colorbar(im, ax=axs[:2], shrink=0.6)
+im = axs[2].pcolormesh(arr, rasterized=True)
+fig.colorbar(im, ax=axs[2], shrink=0.6)
+
+############################################################################
+# The API to make a single-axes behave like a list of axes is to specify
+# it as a list, as below:
```
Comment (afvincent) [inline tutorials/intermediate/constrainedlayout_guide.py]: Similar “remark” about the word "list" (vs. a more general term to mean any kind of sequence) than above. TBH I do not have a clear opinion on this, but just wanted to make sure that one noticed the small subtlety.

File: lib/matplotlib/axes/_subplots.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/axes/_subplots.py]: Ooops, this shouldn't have been in this PR....

**Comment by afvincent** [comment]
Well, I have been too much exposed to Matplotlib and to its internals not to be biased, but `ax=[axs[2]]` is something that I actually tried (before this PR) when I was originally looking for a workaround. It was based on the loose intuition that a single `Axes` instance might be processed differently than a “group” of them. That is why IMO that API is not totally unreasonable ;). (But I would be glad as well if anybody would come suggesting another neat API.)

**Pull request merged**